### PR TITLE
[SolYul] Index access to fixed bytes and byte arrays.

### LINKED ISSUE #9109: [Sol-Yul] Support indexing fixed- or dynamic-length bytes arrays
This should compile:
```
contract C {
    function to_little_endian_64(uint64 value) internal pure returns (bytes memory ret) {
        ret = new bytes(8);
        bytes8 bytesValue = bytes8(value);
        // Byteswapping during copying to bytes.
        ret[0] = bytesValue[7];
        ret[1] = bytesValue[6];
        ret[2] = bytesValue[5];
        ret[3] = bytesValue[4];
        ret[4] = bytesValue[3];
        ret[5] = bytesValue[2];
        ret[6] = bytesValue[1];
        ret[7] = bytesValue[0];
    }
}
```

Both `ret[0]` and `bytesValue[0]` are triggering this assertion:
```
Unimplemented feature:
/Users/alex/Projects/solidity/libsolidity/codegen/ir/IRGeneratorForStatements.cpp(1730): Throw in function virtual void solidity::frontend::IRGeneratorForStatements::endVisit(const solidity::frontend::IndexAccess &)
Dynamic exception type: boost::wrapexcept<solidity::langutil::UnimplementedFeatureError>
std::exception::what: 
[solidity::util::tag_comment*] = 
```

### PR DESCRIPTION
**Pull request by chriseth**
Fixes #9109

### CODE REVIEW

**Review by leonardoalt — CHANGES_REQUESTED** [review]
Test?

**Comment by chriseth** [comment]
Forgot to commit test.

File: libsolidity/codegen/ir/IRGeneratorForStatements.cpp [hunk 1]
```diff
@@ -1763,7 +1763,23 @@ void IRGeneratorForStatements::endVisit(IndexAccess const& _indexAccess)
 		}
 	}
 	else if (baseType.category() == Type::Category::FixedBytes)
-		solUnimplementedAssert(false, "");
+	{
+		FixedBytesType const& fixedBytesType = dynamic_cast<FixedBytesType const&>(baseType);
```
Comment (mijovic) [inline libsolidity/codegen/ir/IRGeneratorForStatements.cpp]:
```suggestion
		auto const& fixedBytesType = dynamic_cast<FixedBytesType const&>(baseType);
```

File: libsolidity/codegen/ir/IRGeneratorForStatements.cpp [hunk 2]
```diff
@@ -1763,7 +1763,23 @@ void IRGeneratorForStatements::endVisit(IndexAccess const& _indexAccess)
...
+		solAssert(_indexAccess.indexExpression(), "Index expression expected.");
+
+		IRVariable index{m_context.newYulVariable(), *TypeProvider::uint256()};
+		define(index, *_indexAccess.indexExpression());
+		m_code << Whiskers(R"(
+			if iszero(lt(<index>, <length>)) { invalid() }
+			let <result> := <shl>(byte(<index>, <array>))
```
Comment (mijovic) [inline libsolidity/codegen/ir/IRGeneratorForStatements.cpp]:
```suggestion
			let <result> := <shl248>(byte(<index>, <array>))
```

File: libsolidity/codegen/ir/IRGeneratorForStatements.cpp [hunk 3]
```diff
@@ -1763,7 +1763,23 @@ void IRGeneratorForStatements::endVisit(IndexAccess const& _indexAccess)
...
+		)")
+		("index", index.name())
+		("length", to_string(fixedBytesType.numBytes()))
+		("array", IRVariable(_indexAccess.baseExpression()).name())
+		("shl", m_utils.shiftLeftFunction(256 - 8))
```
Comment (mijovic) [inline libsolidity/codegen/ir/IRGeneratorForStatements.cpp]:
```suggestion
		("shl248", m_utils.shiftLeftFunction(256 - 8))
```

Comment (mijovic) [reply libsolidity/codegen/ir/IRGeneratorForStatements.cpp]: This needs to be changed as well

Comment (leonardoalt) [reply libsolidity/codegen/ir/IRGeneratorForStatements.cpp]: Fixed

**Review by leonardoalt — APPROVED** [review]

**Pull request merged**